src/abstractinterpretation.jl

```diff
@@ -247,6 +247,17 @@ function CC.abstract_call_method(interp::JETInterpreter, method::Method, @nospec
         end
     end
 
+    if method.name === :iterate
+        v, _, _ = ret
+        ret_type = (isa(v, Const) ? Core.Typeof(v.val) :
+                    isa(v, Core.PartialStruct) ? v.typ :
+                    isa(v, DataType) || isa(v, Union) ? v :
+                    return)::Type
+        if ret_type !== Union{Nothing, Tuple{Int64, Int64}}
```

Comment:
```suggestion
        ret_type = widenconst(v)
        if !(ret_type <: Union{Nothing, Tuple{Int64, Int64}})
```

This is what you want, and this approach doesn't work, since return type of infinite `iterate` will conform to `<: Union{Nothing, Type{Any, Any}}`

Reply:
Thanks for the feedback! Is there a way to assert that the type of the tuple (i.e. `Tuple{Int64, Int64}`) inside of the union is a subtype of `Tuple{Any, Any}`? That way we can still check that the return type is a union of `Nothing` and a tuple.

This is just an idea I had but maybe there's a better solution.

Reply:
Because [Julia's tuple is covariant in their parameters](https://docs.julialang.org/en/v1/manual/types/#Tuple-Types), `t <: Union{Nothing, Tuple{Any, Any}}` is sufficient for that check, no ?
```julia
julia> (1, 2) isa Union{Nothing, Tuple{Any,Any}}
true

julia> nothing isa Union{Nothing, Tuple{Any,Any}}
true

julia> (:iterval, :iterstate) isa Union{Nothing, Tuple{Any,Any}}
true
```
(note that JET is currently designed to be permissive about errors, and so we will use `typeintersect(ret_type, Union{Nothing,Tuple{Any,Any}}) === Bottom` instead)

The problem here is that infinite iteration can only be detected by validating both statement types and control flows inside a local frame.
```julia
julia> struct NeverTerminate
           val::Int
       end

julia> Base.iterate(nv::NeverTerminate, state = 0) = state>nv.val ? (state, state+1) : (state, state+1)

julia> code_typed(; optimize=false) do
           r = 0
           for i in NeverTerminate(100)
               r += 1
           end
           return r
       end |> first
CodeInfo(
1 ─       (r = 0)::Core.Const(0)
│   %2  = Main.NeverTerminate(100)::Core.Const(NeverTerminate(100))
│         (@_2 = Base.iterate(%2))::Core.Const((0, 1))
│   %4  = (@_2::Core.Const((0, 1)) === nothing)::Core.Const(false)
│   %5  = Base.not_int(%4)::Core.Const(true)
└──       %5::Any
2 ┄ %7  = @_2::Tuple{Int64, Int64}
│         (i = Core.getfield(%7, 1))::Int64
│   %9  = Core.getfield(%7, 2)::Int64
│         (r = r + 1)::Int64
│         (@_2 = Base.iterate(%2, %9))::Tuple{Int64, Int64}
│   %12 = (@_2 === nothing)::Core.Const(false)
│   %13 = Base.not_int(%12)::Core.Const(true)
│         %13::Any
└──       goto #2
3 ─       Core.Const(:(return r))::Union{}
) => Union{}
```
So for example we need to report an error by analyzing both `(@_2 = Base.iterate(%2))::Core.Const((0, 1))` and `(@_2 = Base.iterate(%2, %9))::Tuple{Int64, Int64}` come from iteration protocol and both iteration never terminates.
And because of this we will likely to work on the `typeinf_local(::JETInterpreter, ::InferenceState)` overload, where every statement has been annotated.